subprojects/scala/src/main/java/org/gradle/language/scala/tasks/AbstractScalaCompile.java

```diff
@@ -74,18 +74,18 @@
 @DisableCachingByDefault(because = "Abstract super-class, not to be instantiated directly")
 public abstract class AbstractScalaCompile extends AbstractCompile implements HasCompileOptions {
     protected static final Logger LOGGER = Logging.getLogger(AbstractScalaCompile.class);
-    private final BaseScalaCompileOptions scalaCompileOptions;
+    private BaseScalaCompileOptions scalaCompileOptions;
     private final CompileOptions compileOptions;
     private final RegularFileProperty analysisMappingFile;
     private final ConfigurableFileCollection analysisFiles;
     private final Property<JavaLauncher> javaLauncher;
 
-    protected AbstractScalaCompile(BaseScalaCompileOptions scalaCompileOptions) {
+    protected AbstractScalaCompile(Class<? extends BaseScalaCompileOptions> scalaCompileOptionsClass) {
         ObjectFactory objectFactory = getObjectFactory();
         this.analysisMappingFile = objectFactory.fileProperty();
         this.analysisFiles = getProject().files();
         this.compileOptions = objectFactory.newInstance(CompileOptions.class);
-        this.scalaCompileOptions = scalaCompileOptions;
+        this.scalaCompileOptions = getProject().getObjects().newInstance(scalaCompileOptionsClass);
```

Comment:
I don't think we need to pass `ScalaCompileOptions.class` as an argument, we can just inline it here.